### ISSUE TITLE
Type narrowing with `Literal[Enum]` seems to fail on complex code

### ISSUE THREAD

**Issue by BlueGlassBlock** [microsoft/pyright#4661]
**Describe the bug**
Complex `TypeAlias`es and code branches seems to cause pyright fail to narrow down types from a `Union` containing `Literal[Enum]` pattern.

**To Reproduce**
This is a stripped part from a large codebase, so it seems that the code is useless.
Simpler code may be unable to trigger this problem.

> File: `sentinel.py`
```python
import enum
from typing_extensions import TypeAlias
from typing import Literal, Union, TypeVar


class _SentinelProvider(enum.Enum):
    _Sentinel = object()


T = TypeVar("T")

Sentinel = _SentinelProvider._Sentinel

FlagAlias: TypeAlias = Literal[Sentinel]
MaybeFlag: TypeAlias = Union[Literal[Sentinel], T]
```

> File: `util.py`
```python
from __future__ import annotations
from typing import TypeVar, Union, Generic
from weakref import WeakSet
from typing_extensions import reveal_type
from sentinel import MaybeFlag, Sentinel, FlagAlias


class MatchEntry:
    def __init__(self, tokens: list[Union[str, int]]) -> None:
		self.nodes: list[MaybeFlag[frozenset[str]]] = [ # Cannot assign member "nodes" for type "MatchEntry"
            frozenset(token) if isinstance(token, str) else Sentinel for token in tokens
        ]

T_MatchEntry = TypeVar("T_MatchEntry", bound=MatchEntry)


class MatchNode(Generic[T_MatchEntry]):
    __slots__ = ("next", "entries")
    next: dict[MaybeFlag[str], MatchNode[T_MatchEntry]]
    entries: WeakSet[T_MatchEntry]

    def __init__(self) -> None:
        self.next = {}
        self.entries = WeakSet()

    def push(self, entry: T_MatchEntry, index: int = 0) -> None:
        current: Union[frozenset[str], FlagAlias] = entry.nodes[index]
        if current is Sentinel:
            reveal_type(current) # Type of "current" is "frozenset[str] | Literal[_SentinelProvider._Sentinel]"
            self.next.setdefault(current, MatchNode()).push(entry, index + 1)
            """
           Argument of type "MaybeFlag[frozenset[str]]" cannot be assigned to parameter "__key" of type "MaybeFlag[str]" in function "setdefault"
		  Type "MaybeFlag[frozenset[str]]" cannot be assigned to type "MaybeFlag[str]"
		    Type "frozenset[str]" cannot be assigned to type "MaybeFlag[str]"
		      "frozenset[str]" is incompatible with "str"
		      "frozenset[str]" cannot be assigned to type "Literal[_SentinelProvider._Sentinel]"
            """
```

**Expected behavior**
The above code type checks.

**VS Code extension or command-line**
Pylance 2023.2.31

**Comment by erictraut** [comment]
The problem here is that the inferred type of the variable `Sentinel` is `_SentinelProvider`. Your code ie assuming that its inferred type is `Literal[_SentinelProvider._Sentinel]`. You can fix this by annotating the variable `Sentinel` as `Final`.

```python
Sentinel: Final = _SentinelProvider._Sentinel
```

**Comment by BlueGlassBlock** [comment]
The solution didn't solve my problem, the code didn't pass type checking. However, I've investigated some weird behaviors.

![image](https://user-images.githubusercontent.com/50769752/219942116-348502df-016a-44e7-8e7c-e584ea753cf1.png)


You can see that the type of `Sentinel` is incorrect in the other module's methods.

**Comment by BlueGlassBlock** [comment]
Another detail: `Sentinel: Final[Literal[_SentinelProvider._Sentinel]] = _SentinelProvider._Sentinel` doesn't work.

**Comment by erictraut** [comment]
Yes, you're right. When I reproed the problem, I moved all of the code into a single file. There was a bug in the logic that applied only in cases where a `Final` variable was imported from another file. The literal value was being stripped from the type.

The fix will be included in the next release.
